packages/pyright-internal/src/analyzer/service.ts

```diff
@@ -600,71 +600,54 @@ export class AnalyzerService {
-            );
-            configOptions.pythonPath = this.fs.realCasePath(
-                Uri.file(commandLineOptions.pythonPath, this.serviceProvider, /* checkRelative */ true)
-            );
-        }
-
-        if (commandLineOptions.pythonEnvironmentName) {
-            this._console.info(
-                `Setting environmentName for service "${this._instanceName}": ` +
-                    `"${commandLineOptions.pythonEnvironmentName}"`
-            );
-            configOptions.pythonEnvironmentName = commandLineOptions.pythonEnvironmentName;
-        }
-
-        // The pythonPlatform and pythonVersion from the command-line can be overridden
-        // by the config file, so initialize them upfront.
-        configOptions.defaultPythonPlatform = commandLineOptions.pythonPlatform;
-        configOptions.defaultPythonVersion = commandLineOptions.pythonVersion;
-        configOptions.ensureDefaultExtraPaths(
-            this.fs,
-            commandLineOptions.autoSearchPaths ?? false,
-            commandLineOptions.extraPaths
-        );
-
-        commandLineOptions.includeFileSpecs.forEach((fileSpec) => {
-            configOptions.include.push(getFileSpec(projectRoot, fileSpec));
-        });
-
-        commandLineOptions.excludeFileSpecs.forEach((fileSpec) => {
-            configOptions.exclude.push(getFileSpec(projectRoot, fileSpec));
-        });
-
-        commandLineOptions.ignoreFileSpecs.forEach((fileSpec) => {
-            configOptions.ignore.push(getFileSpec(projectRoot, fileSpec));
-        });
-
-        configOptions.disableTaggedHints = !!commandLineOptions.disableTaggedHints;
-
-        configOptions.initializeTypeCheckingMode(commandLineOptions.typeCheckingMode ?? 'standard');
 
+        // If we found a config file, load it and apply its settings.
         const configs = this._getExtendedConfigurations(configFilePath ?? pyprojectFilePath);
-
         if (configs && configs.length > 0) {
+            // With a pyrightconfig.json set, we want the typeCheckingMode to always be standard
+            // as that's what the Pyright CLI will expect. Command line options (if not a VS code extension) and
+            // the config file can override this.
+            configOptions.initializeTypeCheckingMode('standard');
+
+            // Then we apply the config file settings. This can update the
+            // the typeCheckingMode.
             for (const config of configs) {
                 configOptions.initializeFromJson(
                     config.configFileJsonObj,
                     config.configFileDirUri,
                     this.serviceProvider,
-                    host,
-                    commandLineOptions
+                    host
                 );
             }
+
+            // When not in VS code mode, command line options override config file options.
+            if (!commandLineOptions.fromVsCodeExtension) {
```

Comment:
It occurs to me that "fromVsCodeExtension" would be more accurately called "fromLanguageServer" since it also applies when using the pyright LSP. We can change that in a subsequent PR if you don't want to do it here.

The comments below also refer to "VS code extension", and they should more accurately refer to "language server".

Reply:
Good idea. Thanks. I'll change it.